aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -257,12 +257,11 @@ Tensor prod(const Tensor &self) {
 static inline Tensor &mean_out(Tensor &result, const Tensor &self, IntArrayRef dim,
                  bool keepdim, optional<ScalarType> opt_dtype) {
   ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
-  AT_CHECK(
-      at::isFloatingType(scalarType),
-      "Can only calculate the mean of floating types. Got ",
-      toString(scalarType),
-      " instead.");
   ScalarType dtype = get_dtype(result, self, opt_dtype, true);
+
+  if (at::isIntegralType(scalarType) && !opt_dtype.has_value()){
+    dtype = ScalarType::Double;
```

Comment:
nah we shouldn't use double as default. Instead, we should use whatever the default floating point type it is.

Reply:
Oh right. Can you please guide me regarding the same as I don't know how to get it. Thank You

Reply:
@SsnL would you block landing until this is fixed?

To get the default floating point type, it can be retrieved via `getDefaultTensorOptions()`

Reply:
@ezyang Thank You for the guidance.

```C++
ScalarType dtype = c10::typeMetaToScalarType(c10::get_default_dtype());
```

Will this have the same result as using
```C++
auto options = c10::getDefaultTensorOptions();
ScalarType dtype = c10::typeMetaToScalarType(options.dtype());
```

Both result to `float32` on the system that I use.

If yes, which one is preferable?

Reply:
@ezyang I think we should fix this before landing.  Otherwise it still needs to be fixed before a release to avoid nasty BC problems in future.

Reply:
@SsnL @ezyang I have made the changes. Please have a look to see if it was what you were looking for.

As mentioned above, we can use
`ScalarType dtype = c10::typeMetaToScalarType(c10::get_default_dtype());`
without having to touch the `ATen.h` file.

Reply:
Thanks! Could you add a test that the returned value matches the default dtype (say, by checking outputs with default dtype being double and float).

Reply:
Sure. Will give it a go.

Reply:
@SsnL I have added a test. Could you please take a look to see if it is similar to what you expected?

Reply:
Yep that looks correct. Thanks!

Reply:
@SsnL The implementation with 
```
ScalarType dtype = c10::typeMetaToScalarType(c10::getDefaultTensorOptions().dtype());
```

Failed in many (roughly 20) tests with, 
`torch.float32 is not torch.float64`.

However, with
```
ScalarType dtype = c10::typeMetaToScalarType(c10::get_default_dtype());
```
None of the failing tests are due to this.
For the 5 failings test, 
3 MacOs test failed during setting up of the environment.
1 Lint Test surprisingly failed on a file untouched by this PR. 
1 failed on Windows Platfrom in Dataloader Test with `Out of Memory`.

So I assume, [`c10::get_default_dtype()`](https://github.com/pytorch/pytorch/blob/master/c10/core/DefaultDtype.cpp) tracks the `dtype` set using `torch.set_default_dtype`, while from what I understand, [`c10::getDefaultTensorOptions()`](https://github.com/pytorch/pytorch/blob/master/c10/core/DefaultTensorOptions.h)  always returns the system default floating point.

So I believe, we should use, 
```
ScalarType dtype = c10::typeMetaToScalarType(c10::get_default_dtype());
```

Please let me know about your view and if I was wrong somewhere.
Thank You.